File: Digito Verificador/Modulo 10/DV Verificador/verifdig_mod10.pas
program ValidarMatricula;

uses
  SysUtils;

function ValidarMatricula(const MATRICULA: string): Boolean;
begin
  if Length(MATRICULA) <> 9 then
    Exit(False);
  if not (MATRICULA[1] in ['0'..'9']) or not (MATRICULA[2] in ['0'..'9']) or not (MATRICULA[3] in ['0'..'9']) then
    Exit(False);
  if MATRICULA[4] <> '.' then
    Exit(False);
  if not (MATRICULA[5] in ['0'..'9']) or not (MATRICULA[6] in ['0'..'9']) or not (MATRICULA[7] in ['0'..'9']) then
    Exit(False);
  if MATRICULA[8] <> '-' then
    Exit(False);
  if not (MATRICULA[9] in ['0'..'9']) then
    Exit(False);
  ValidarMatricula := True;
end;

procedure RemoverCaractere(var CADEIA: string; CARACTERE: Char);
var
  POS1, POS2: Integer;
begin
  POS1 := 1;
  POS2 := 1;
  while POS1 <= Length(CADEIA) do
  begin
    if CADEIA[POS1] <> CARACTERE then
    begin
      CADEIA[POS2] := CADEIA[POS1];
      POS2 := POS2 + 1;
    end;
    POS1 := POS1 + 1;
  end;
  SetLength(CADEIA, POS2 - 1);
end;

function Mod10Check(const MATRICULA: string): Integer;
var
  MATRIC_INT: array[1..6] of Integer;
  SOMA, I, DV: Integer;
begin
  for I := 1 to 6 do
    MATRIC_INT[I] := StrToIntDef(MATRICULA[I], 0);
  SOMA := 0;
  for I := 1 to 6 do
  begin
    if (I mod 2 = 0) then
    begin
      if MATRIC_INT[I] * 2 > 9 then
        SOMA := SOMA + (MATRIC_INT[I] * 2 - 9)
      else
        SOMA := SOMA + (MATRIC_INT[I] * 2);
    end
    else
      SOMA := SOMA + MATRIC_INT[I];
  end;
  DV := (10 - (SOMA mod 10)) mod 10;
  Mod10Check := DV;
end;

var
  MATRIC_CPT, MATRIC_SPT: string;
  MATRIC_INT: array[1..7] of Integer;
  I, DVC, DVE: Integer;

begin
  repeat
    Write('Informe a matricula no formato 999.999-9: ');
    ReadLn(MATRIC_CPT);
    if not ValidarMatricula(MATRIC_CPT) then
    begin
      WriteLn;
      WriteLn('Entrada invalida, tente novamente.');
      WriteLn; WriteLn;
    end;
  until ValidarMatricula(MATRIC_CPT);

  MATRIC_SPT := MATRIC_CPT;
  RemoverCaractere(MATRIC_SPT, '.');
  RemoverCaractere(MATRIC_SPT, '-');
  DVC := Mod10Check(MATRIC_SPT);

  for I := 1 to 7 do
    MATRIC_INT[I] := StrToIntDef(MATRIC_SPT[I], 0);
  DVE := MATRIC_INT[7];

  WriteLn;
  if DVC = DVE then
    WriteLn('Matricula valida.')
  else
    WriteLn('Matricula invalida.');
  WriteLn;
  
  Write('Pressione <Enter> para encerrar... ');
  ReadLn;
end.

File: Digito Verificador/Modulo 10 Padrao/DV Verificador/verifdig_mod10.java
// ValidarMatricula
import java.util.Scanner;

public class Main {
    public static boolean validarMatricula(String MATRICULA) {
        if (MATRICULA.length() != 9)
            return false;
        if (!(Character.isDigit(MATRICULA.charAt(0))) || !(Character.isDigit(MATRICULA.charAt(1))) || !(Character.isDigit(MATRICULA.charAt(2))))
            return false;
        if (MATRICULA.charAt(3) != '.')
            return false;
        if (!(Character.isDigit(MATRICULA.charAt(4))) || !(Character.isDigit(MATRICULA.charAt(5))) || !(Character.isDigit(MATRICULA.charAt(6))))
            return false;
        if (MATRICULA.charAt(7) != '-')
            return false;
        if (!(Character.isDigit(MATRICULA.charAt(8))))
            return false;
        return true;
    }

    public static String removerCaractere(String CADEIA, char CARACTERE) {
        return CADEIA.replace(Character.toString(CARACTERE), "");
    }

    public static int mod10Check(String MATRICULA) {
        int[] MATRIC_INT = new int[6];
        int SOMA = 0;
        int DV = 0;
        for (int i = 0; i < 6; i++)
            MATRIC_INT[i] = Integer.parseInt(Character.toString(MATRICULA.charAt(i)));
        for (int i = 0; i < 6; i++) {
            if ((i + 1) % 2 == 0) {
                if (MATRIC_INT[i] * 2 > 9)
                    SOMA += MATRIC_INT[i] * 2 - 9;
                else
                    SOMA += MATRIC_INT[i] * 2;
            } else {
                SOMA += MATRIC_INT[i];
            }
        }
        DV = (10 - (SOMA % 10)) % 10;
        return DV;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String MATRIC_CPT = "";
        String MATRIC_SPT = "";
        int[] MATRIC_INT = new int[7];
        int DVC = 0;
        int DVE = 0;

        do {
            System.out.print("Informe a matricula no formato 999.999-9: ");
            MATRIC_CPT = scanner.nextLine();

            if (!validarMatricula(MATRIC_CPT)) {
                System.out.println();
                System.out.println("Entrada invalida, tente novamente.");
                System.out.println();
                System.out.println();
            }
        } while (!validarMatricula(MATRIC_CPT));

        MATRIC_SPT = MATRIC_CPT;
        MATRIC_SPT = removerCaractere(MATRIC_SPT, '.');
        MATRIC_SPT = removerCaractere(MATRIC_SPT, '-');
        DVC = mod10Check(MATRIC_SPT);

        for (int i = 0; i < 7; i++)
            MATRIC_INT[i] = Integer.parseInt(Character.toString(MATRIC_SPT.charAt(i)));
        DVE = MATRIC_INT[6];

        System.out.println();
        if (DVC == DVE)
            System.out.println("Matricula valida.");
        else
            System.out.println("Matricula invalida.");
        System.out.println();

        System.out.print("Pressione <Enter> para encerrar... ");
        scanner.nextLine();
    }
}
